src/v/cluster/tm_stm.cc

```diff
@@ -102,6 +104,75 @@ ss::future<checked<model::term_id, tm_stm::op_status>> tm_stm::barrier() {
         });
 }
 
+ss::future<> tm_stm::checkpoint_in_memory_state() {
+    iobuf key;
+    reflection::serialize(key, _insync_term);
+    iobuf value;
+    co_await reflection::async_adl<mem_txes_type>{}.to(value, _mem_txes);
```

Comment:
making a snapshot is a too invasive solution, there is a simpler way:
  - introduce new flag for a transaction e.g. `transferring`
  - on a partition movement:
     - take a write lock to have global owership
     - for all ongoing transaction:
        - set `transferring` to true
        - save it to log
     - step down or reject all new write requests

on the receiver side when processing a transaction: if a `transferring` is set to true - reset it and bump the etag.

Reply:
> save it to log

Thats interesting.. perhaps I'm missing something.. don't we need to preserve `_mem_txes` state as is.? We have code like below in multiple places where we try to look up state in `mem_txes` and throw if a tx is not found.

```
........
    auto ptx = _mem_txes.find(tx_id);
    if (ptx == _mem_txes.end()) {
        co_return tm_stm::op_status::not_found;
    }
 ```

saving it to the log means that we lose the mem_state? because on the receiver we populate the `log_txes` state for these `transferring` transactions.. or you mean we need to reconstruct the `mem_txes` for all these `transferring` transactions on assuming leadership but in that case the code is not too different from this patch (probably even mroe complicated because we need to undo log_txes changes), so a bit confused.

Reply:
If we replace the txn access from `_mem_txes` to `get_tx` then we may ignore the memory state during the leadership transferring.

**mark_tx_preparing**

```
// from

auto ptx = _mem_txes.find(tx_id);
if (ptx == _mem_txes.end()) {
    co_return tm_stm::op_status::not_found;
}

// to

auto ptx = get_tx(tx_id);
if (!ptx.has_value()) {
    co_return tm_stm::op_status::not_found;
}
if (ptx.value().etag != expected_term) {
    co_return tm_stm::op_status::unknown;
}
```

**mark_tx_aborting**

```
// from

auto ptx = _mem_txes.find(tx_id);
if (ptx == _mem_txes.end()) {
    co_return tm_stm::op_status::not_found;
}
auto tx = ptx->second;

// to

auto ptx = get_tx(tx_id);
if (!ptx.has_value()) {
    co_return tm_stm::op_status::not_found;
}
if (ptx.value().etag != expected_term) {
    co_return tm_stm::op_status::unknown;
}
auto tx = ptx.value();
```

**add_partitions**

```
// from (it checks freshness of the txn before invoking add_partitions)

auto ptx = _mem_txes.find(tx_id);
if (ptx == _mem_txes.end()) {
    co_return tm_stm::op_status::unknown;
}

// to

auto ptx = get_tx(tx_id);
if (!ptx.has_value()) {
    co_return tm_stm::op_status::not_found;
}
```

**add_group**

```
// from (it checks freshness of the txn before invoking add_group)

auto ptx = _mem_txes.find(tx_id);
if (ptx == _mem_txes.end()) {
    co_return tm_stm::op_status::unknown;
}

// to

auto ptx = get_tx(tx_id);
if (!ptx.has_value()) {
    co_return tm_stm::op_status::not_found;
}
```